pennylane_forest/qc.py

```diff
@@ -0,0 +1,407 @@
+        ro = self.prog.declare("ro", "BIT", len(qubits))
+        for i, q in enumerate(qubits):
+            self.prog.inst(MEASURE(q, ro[i]))
+
+        self.prog.wrap_in_numshots_loop(self.shots)
+
+    def apply_parametric_operations(self, operations):
+        """Applies a parametric program by applying parametric operation with symbolic parameters.
+
+        Args:
+            operations (List[pennylane.Operation]): quantum operations that need to be applied
+
+        Returns:
+            pyquil.Prgram(): a pyQuil Program with the given operations
+        """
+        prog = Program()
+        # Apply the circuit operations
+        for i, operation in enumerate(operations):
+            # map the operation wires to the physical device qubits
+            device_wires = self.map_wires(operation.wires)
+
+            if i > 0 and operation.name in ("QubitStateVector", "BasisState"):
+                raise DeviceError(
+                    "Operation {} cannot be used after other Operations have already been applied "
+                    "on a {} device.".format(operation.name, self.short_name)
+                )
+
+            # Prepare for parametric compilation
+            par = []
+            for param in operation.data:
+                if getattr(param, "requires_grad", False) and operation.name != "BasisState":
+                    # Using the idx for trainable parameter objects to specify the
+                    # corresponding symbolic parameter
+                    parameter_string = "theta" + str(id(param))
+
+                    if parameter_string not in self._parameter_reference_map:
+                        # Create a new PyQuil memory reference and store it in the
+                        # parameter reference map if it was not done so already
+                        current_ref = self.prog.declare(parameter_string, "REAL")
+                        self._parameter_reference_map[parameter_string] = current_ref
+
+                    # Store the numeric value bound to the symbolic parameter
+                    self._parameter_map[parameter_string] = [param.unwrap()]
+
+                    # Appending the parameter reference to the parameters
+                    # of the corresponding operation
+                    par.append(self._parameter_reference_map[parameter_string])
+                else:
+                    par.append(param)
+
+            prog += self._operation_map[operation.name](*par, *device_wires.labels)
+
+        return prog
+
+    def execute(self, circuit, **kwargs):
+        if self.parametric_compilation:
+            self._circuit_hash = circuit.graph.hash
+        return super().execute(circuit, **kwargs)
+
+    def generate_samples(self):
+        if self.parametric_compilation:
+            for region, value in self._parameter_map.items():
+                self.prog.write_memory(region_name=region, value=value)
+
+        if "pyqvm" in self.qc.name:
```

Comment:
This might not be necessary anymore, but I want to be able to run the pyqvm test suite before deleting it (along with moving the special case of `pyqvm`s in `apply` to the `QVMDevice`)